File: Sources/FocusCore/FocusFeature.swift
import Foundation

public struct FocusFeature: Codable, Equatable, Sendable {
    public enum Action: String, Codable, Sendable {
        case prune
        case repair
    }

    public enum SettingKey: String, Codable, Sendable {
        case redirectEnabled
        case playerMaskEnabled
        case searchMaskEnabled
        case dynamicMaskEnabled
        case debugMode
    }

    public let featureId: String
    public let requiredSelectors: [String]
    public let optionalSelectors: [String]
    public let action: Action
    public let css: String
    public let script: String?
    public let settingKey: SettingKey?

    public init(
        featureId: String,
        requiredSelectors: [String] = [],
        optionalSelectors: [String] = [],
        action: Action,
        css: String,
        script: String? = nil,
        settingKey: SettingKey? = nil
    ) {
        self.featureId = featureId
        self.requiredSelectors = requiredSelectors
        self.optionalSelectors = optionalSelectors
        self.action = action
        self.css = css
        self.script = script
        self.settingKey = settingKey
    }
}

File: focus-core/src/commonMain/kotlin/org/bilibilifocus/core/model/FocusFeature.kt
package org.bilibilifocus.core.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class FocusFeature(
    val featureId: String,
    val requiredSelectors: List<String> = emptyList(),
    val optionalSelectors: List<String> = emptyList(),
    val action: Action,
    val css: String,
    val script: String? = null,
    val settingKey: SettingKey? = null,
) {
    @Serializable
    enum class Action {
        @SerialName("prune") PRUNE,
        @SerialName("repair") REPAIR,
    }

    @Serializable
    enum class SettingKey {
        @SerialName("redirectEnabled") REDIRECT_ENABLED,
        @SerialName("playerMaskEnabled") PLAYER_MASK_ENABLED,
        @SerialName("searchMaskEnabled") SEARCH_MASK_ENABLED,
        @SerialName("dynamicMaskEnabled") DYNAMIC_MASK_ENABLED,
        @SerialName("debugMode") DEBUG_MODE,
    }
}
